nilearn/_utils/bids.py

```diff
@@ -0,0 +1,61 @@
+"""BIDS related utilities."""
+
+from typing import Any
+
+
+def entities() -> dict[str, list[str]]:
+    """Return a dictionary of BIDS entities.
+
+    Entities are listed in the order they should appear in a filename.
+
+    https://bids-specification.readthedocs.io/en/stable/appendices/entities.html # noqa
+
+    Note that:
+
+    - this only contains the entities for functional data
+
+    https://github.com/bids-standard/bids-specification/blob/master/src/schema/rules/files/raw/func.yaml#L13 # noqa
+    https://github.com/bids-standard/bids-specification/blob/master/src/schema/rules/files/deriv/imaging.yaml#L29 # noqa
+
+    Returns
+    -------
+    Dictionary of raw and derivatives entities : dict[str, list[str]]
+
+    """
+    return {
+        "raw": [
+            "sub",
+            "ses",
+            "task",
+            "acq",
+            "ce",
+            "rec",
+            "dir",
+            "run",
+            "echo",
+            "part",
+        ],
+        "derivatives": ["space", "res", "den", "desc"],
+    }
+
+
+def validate_label(label: Any) -> None:
```

Comment:
Why annotate it as `Any` if it should be `str`?

Reply:
Well I would say that the point of this validation is ensuring the label is a string so I would not put any constraints on the input type.

Reply:
I fathom what you mean, but that's not what annotations are envisioned for.

By design, annotations are there to inform users about what kind of input is expected. Like mentioning them in docstrings, but more powerful & versatile than just for human perusal.

Developers can choose to add validation to ensure the program does not continue in case of inadvertent input type, whether by a user or generated by some step in the program.

To indicate to the user via annotations that any input type is acceptable only to raise an error for every non-string input does not serve any purpose. It would be like stating in the docstring that any value is acceptable, except every other type terminates the program.

Since labels are expected to be string, the annotation should be `str`.

Also, `validate_labels` does not tell the code reader anything about the validation itself. It is too generic and nondescriptive.  
I suggest `validate_label_as_alphanumeric` or `verify_label_is_alphanumeric`.

Reply:
Also, `validate_labels` does not tell the code reader anything about the validation itself. It is too generic and nondescriptive.  
I suggest `validate_label_as_alphanumeric` or `verify_label_is_alphanumeric`.

Reply:
I would say that within the bids module validate_label has a very clear meaning: label have a very clear definition within BIDS, hence why I added the link to the relevant part of the spec.

Reply:
You are not incorrect,  and we may very well leave it like this and most people will not be affected much.

I would like to offer an additional perspective; making it more explicit without making it too long, significantly reduces the cognitive load when reading error traces, and logs. It also makes small individual but significant in-totality contributions in easing the onboarding for new people people unfamiliar with BIDS & fMRI (as I was when I joined the Nilearn team) 

Validation functions are better if their names also indicate the validation being performed. For example, `label` can mean a specific thing in BIDS, but it is also a very general word, used within the same context, but as *A* label, instead of *the* label. A hypothetical example would be that `validate_label` may also mean that you are checking from a list of valid labels, but here it is only checking if its string and alphanumeric.

Anyway, that's my perspective. Up to you; I won't block anything over this.

Reply:
> If there is only a single parameter, and it is being annotated as `Any` then the annotation is not adding any value, and should be removed.
> 
> (I removed the long-ass discussion I had typed. I was not sure I was adding any value with that.)

I tend to use type checkers in strict mode and require type annotation for all functions. Hence why I added it here.
Though I believe we are both wasting our times arguing about it here: best keep our energy for to come to an agreement with all maintainers, agree on some guidelines and then move on.

So let's keep thing more liberal for now and remove this.

Feel free to point out any other type annotation you find "too strict" (or just plain weird 😄) and I will back off.

Reply:
> You are not incorrect, and we may very well leave it like this and most people will not be affected much.
> 
> I would like to offer an additional perspective; making it more explicit without making it too long, significantly reduces the cognitive load when reading error traces, and logs. It also makes small individual but significant in-totality contributions in easing the onboarding for new people people unfamiliar with BIDS & fMRI (as I was when I joined the Nilearn team)
> 
> Validation functions are better if their names also indicate the validation being performed. For example, `label` can mean a specific thing in BIDS, but it is also a very general word, used within the same context, but as _A_ label, instead of _the_ label. A hypothetical example would be that `validate_label` may also mean that you are checking from a list of valid labels, but here it is only checking if its string and alphanumeric.
> 
> Anyway, that's my perspective. Up to you; I won't block anything over this.

fair enough

if anyone does not import the whole module but only the function, the function might not be clear enough, so maybe indeed better make the name clearer.

I suspect this should apply to the `entities` function too, no?